alf/utils/nest_utils.py

```diff
@@ -33,179 +31,128 @@ def is_namedtuple(value):
+        "Nest {} must be a dict or namedtuple!".format(nest)
+    fields = nest.keys() if isinstance(nest, dict) else nest._fields
+    for field in fields:
+        value = nest[field] if isinstance(nest, dict) else getattr(nest, field)
+        yield field, value
+
+
+def is_nested(value):
+    """Returns true if the input is one of: `list`, `unnamedtuple`, `dict`, or
+    `namedtuple`. Note that this definition is different from tf's is_nested
+    where all types that are collections.abc.Sequence are defined to be nested.
     """
-    # TF may wrap tuple, which causes construction of tuple to fail.
-    # So we use the original object instead.
-    if isinstance(nest, wrapt.ObjectProxy):
-        nest = nest.__wrapped__
-    if isinstance(example, wrapt.ObjectProxy):
-        example = example.__wrapped__
-    if isinstance(nest, tuple):
-        new_nest = tuple(
-            nest_tuple_to_list(nst, exp) for nst, exp in zip(nest, example))
-        if is_namedtuple(example):
-            # example is a namedtuple
-            new_nest = type(example)(*new_nest)
-        elif isinstance(example, list):
-            # type(example) might be ListWrapper
-            new_nest = type(example)(list(new_nest))
-        return new_nest
-    elif isinstance(nest, list):
-        raise ValueError("list is not expected in nest %s" % nest)
-    elif isinstance(nest, dict):
-        new_nest = {}
-        for k, v in nest.items():
-            new_nest[k] = nest_tuple_to_list(v, example[k])
-        return new_nest
+    return isinstance(value, (list, tuple, dict))
+
+
+def flatten(nest):
+    """Returns a flat list from a given nested structure."""
+    if not is_nested(nest):
+        # any other data type will be returned as it is
+        return [nest]
+    flattened = []
+    if isinstance(nest, list) or is_unnamedtuple(nest):
+        for value in nest:
+            flattened.extend(flatten(value))
     else:
-        return nest
-
-
-def get_nest_batch_size(nest, dtype=None):
-    """Get the batch_size of a nest."""
-    batch_size = tf.shape(tf.nest.flatten(nest)[0])[0]
-    if dtype is not None:
-        batch_size = tf.cast(batch_size, dtype)
-    return batch_size
-
-
-def to_distribution_param_spec(nest):
-    def _to_param_spec(spec):
-        if isinstance(spec, DistributionSpec):
-            return spec.input_params_spec
-        elif isinstance(spec, tf.TensorSpec):
-            return spec
+        for _, value in extract_fields_from_nest(nest):
+            flattened.extend(flatten(value))
+    return flattened
+
+
+def assert_same_structure(nest1, nest2):
+    """Asserts that two structures are nested in the same way."""
+    # When neither is nested, the assertion won't fail
+    if is_nested(nest1) or is_nested(nest2):
+        assert type(nest1) == type(nest2)
+        if isinstance(nest1, list) or is_unnamedtuple(nest1):
+            for value1, value2 in zip(nest1, nest2):
```

Comment:
need to check that nest1 and nest2 have same size. zip will not report error. Missing some unittest cases.

Reply:
OK, python2 did report errors for this. Maybe things have changed in 3.